### ISSUE TITLE
Feature Request: Add an option to not "snap" strokes to their correct position

### ISSUE THREAD

**Issue by martinetd** [SYtor/Kanji-Dojo#100]
Hi, and thanks for this great app!

I've just discovered it by chance (latest updates banner on f-droid site...), and will be trying it out for a while before spending much time fiddling with some details (I have a few pet peeves -- would like to have a way to make the drawing box smaller as I'm lazy and writing on paper is so much smaller than phone screen I cannot relate to it at all; have a way of editing the main translation (or add a custom note) as there are a few wordings I'm not used to/would like to differentiate two meanings further... I'll be happy to do all that work and submit PRs once I'm sure I'll keep using this for a while)

Now for the feature request at hand; rationale first: my brain is lazy and very visual memory-oriented, so for complex kanjis I'll sometimes be able to half-guess the start, and seeing it properly snapped to position will make me recognize it and remember the rest from the clean shape instead of "properly" remembering it.
I can obviously still flag that as bad, but it's a feature I found very little kanji training apps have and I think it makes an appreciable difference for review.

I'd find it great to just keep the strokes as I wrote them instead of cleaning them up one at a time as the kanji is being written.

If it was just about changing what's displayed I'm sure I could kludge something myself, but I think that this kind of change would require some adjustment on the recognition side, because for example it'd need to recognize a kanji that's just shifted to the side or similar transforms...
A possible solution would be to not bother with such adjustments but in this mode ""just"" plug in a completely different algorithm like feeding the strokes to https://f-droid.org/packages/ch.seto.kanjirecog/ and check the kanji that comes up as first match is correct
(although to be fair I might be happy with self evaluation here, being told it's not recognizable has value... Perhaps just enforcing stroke count at start.)

Anyway, technical details can be discussed later -- first, how do you feel about such a mode?

**Comment by sl08154711** [comment]
> for complex kanjis I'll sometimes be able to half-guess the start, and seeing it properly snapped to position will make me recognize it and remember the rest from the clean shape instead of "properly" remembering it.

Yes, I know. Same here. Displaying user submitted strokes instead of the correct stroke, somehow storing all pairs of correct and user drawn strokes and at the end comparing all stroke pairs should be doable. BUT: how would the app know when you are done drawing? Can't go by stroke count because the user might have picked the wrong kanji. So you'd need to tap a button to let the app know. And what do you do if you made a mistake? There should be additional buttons like "clear last stroke" and "clear (all)". Where would they fit?

And when is a character considered to be correct? If all strokes are correct (like it is now) or are you allowed to make one or two mistakes? 

And you are depriving yourself of a learning experience because at the end you would just see the strokes you have drawn and can't see the "correct" way of drawing the character. Should both be overlayed at the end? Might be a visual mess.

Don't get me wrong, I'm all for it. I might even give it a try to actually code it but we all should first talk about expectations.

**Comment by martinetd** [comment]
Yes to most of that -- it'll require a way to undo strokes, and need a button to manually confirm you're done.
I also think it needs to display the correct character after running the check (if only for manual validation) -- in case of mistakes if the algorithm still validates one stroke at a time you could even make differentiate strokes that were correct to the ones that weren't.

I probably shouldn't show this but... the app that spoiled me with this is Kanji Study, so here's a quick capture of how it does it:
https://github.com/syt0r/Kanji-Dojo/assets/1729331/24595109-d97c-48d6-a177-13f172b30d9d
(It's got its share of problems though (biggest of which is no longer working on degoogled phones and not being open), so I was just looking for something else when I saw Kanji Dojo)


I don't think we should aim for 100% the same; if you really care there's more I didn't capture (replaying side by side what you draw vs. ideal stroke order being drawn one at a time etc) but I think that's all polish that can come later.

For starters just having a back/validate button and showing the ideal form at the end would be more than enough to use for me.

**Comment by syt0r** [comment]
Hmm, looks interesting, I think it's worth implementing, maybe I'll call it hard mode 😆

**Comment by syt0r** [comment]
Available in version 2.0.4

**Comment by martinetd** [comment]
Awesome! Thanks a lot!

I've just updated and gave it a try, it's hard mode alright!

Input-wise I see no problem, I can mostly draw what I want.
Just a small nitpick here, drawing 里 it's a bit difficult to draw the bottom bar on my setup (ignore the requested kanji here, just restarted a session to reproduce), because any stroke starting from the bottom left will be ignored:
![kanji_dojo_free_draw](https://github.com/syt0r/Kanji-Dojo/assets/1729331/f55c785b-54c7-4f98-8ffa-a01389cf9217)
(But I think that's mostly a problem I brought on myself making the drawing area smaller...)

Ideally perhaps it would be possible to differentiate a tap (trying to hit the back button) from a stroke (moving away from a corner), but given the check is manual it doesn't matter too much...

On a similar vein review wise, the good/bad buttons mask part of the kanji so it's sometimes hard to check, e.g. does the bottom right last stroke of 読 curl up or not here (but likewise, that's half brought on myself -- I'll try to move the buttons to the right out of the drawing board..)?
![kanji_dojo_draw_confirm](https://github.com/syt0r/Kanji-Dojo/assets/1729331/183cc74f-41fa-4121-9781-9e9e32494241)


Still on review, it's nice to see the final form when validating, that helps correct oneself (pretty much necessary when I had a blank :grin:), but I've had a few times where the detection said I made a mistake and I wasn't totally sure if it was just too strict or if I really messed up, perhaps tapping the kanji to alternate between the correct form and what the user input would make sense?
Being greedy here, there's a few cases where I wasn't sure of the drawing order, so a button to animate it like we have in the dictionary screen would be nice and could make sense even without free drawing (tapping on top kanji here is drawing in progress)
![kanji_dojo_review_screen](https://github.com/syt0r/Kanji-Dojo/assets/1729331/606c6a39-9af7-4d63-95d4-400ea617aa3e)
